jupyter_events/logger.py

```diff
@@ -147,20 +295,38 @@ def emit(self, schema_id: str, version: int, data: dict, timestamp_override=None
             The recorded event data
         """
         # If no handlers are routing these events, there's no need to proceed.
-        if not self.handlers:
+        if (
+            not self.handlers
+            and not self.modified_listeners
+            and not self.unmodified_listeners
+        ):
             return
 
         # If the schema hasn't been registered, raise a warning to make sure
         # this was intended.
         if (schema_id, version) not in self.schemas:
             warnings.warn(
-                f"({schema_id}, {version}) has not been registered yet. If "
-                "this was not intentional, please register the schema using the "
+                f"({schema_id}, {version}) has not "
+                "been registered yet. If this was not intentional, "
+                "please register the schema using the "
                 "`register_event_schema` method.",
                 SchemaNotRegistered,
             )
             return
 
+        # Validate the raw event data.
+        self.schemas.validate_event(schema_id, version, data)
+
+        # Deep copy the data and modify the copy.
+        modified_data = copy.deepcopy(data)
```

Comment:
This came up in contributor hour, it seems like a good optimization here is to check whether anyone has added `unmodified_listeners` and if not, only validate once and don't create the deep copy.

Reply:
Thanks! Fixed!